File: problems_list/0103.二叉树的锯齿形层序遍历/0103-二叉树的锯齿形层序遍历.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function(root) {
    if(root===null) return [];
    let queue=[root], odd=0, ans=[];
    while(queue.length){
        let layer=[], n=queue.length;
        for(let i=0; i<n; i++){
            let node=queue.shift();
            if(node.left) queue.push(node.left);
            if(node.right) queue.push(node.right);
            layer.push(node.val);
        }
        if(odd%2) layer.reverse();
        odd++;
        ans.push(layer);
    }
    return ans;
};

File: problems_list/0103.二叉树的锯齿形层序遍历/0103-二叉树的锯齿形层序遍历.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if(root==nullptr) return {};
        vector<vector<int>> ans;
        queue<TreeNode*> q;
        q.push(root);

        int layer=0;
        while(!q.empty()){
            int len=q.size();
            vector<int> tmp;
            for(int i=0; i<len; i++){
                TreeNode* node=q.front();
                q.pop();
                tmp.push_back(node->val);
                if(node->left) q.push(node->left);
                if(node->right) q.push(node->right);
            }
            if(layer%2) reverse(tmp.begin(), tmp.end());
            ans.push_back(tmp);
            layer++;
        }
        return ans;
    }
};